src/librustdoc/html/render.rs

```diff
@@ -854,6 +853,35 @@ fn write_shared(cx: &Context,
     }
     try_err!(writeln!(&mut w, "initSearch(searchIndex);"), &dst);
 
+    // Create the directory structure ahead of time
+    let mut item = krate.module.clone().unwrap();
+    item.name = Some(krate.name.clone());
+    trace!("Creating directory tree for {:#?}", item);
+
+    fn recurse(dst: &Path, item: &clean::Item) -> Result<(), Error> {
+        if !item.is_mod() { return Ok(()); }
+        let name = item.name.as_ref().unwrap();
+        let dst = dst.join(name);
+
+        debug!("creating directory tree: recursing into {}", name);
+
+        try_err!(fs::create_dir_all(&dst), &dst);
```

Comment:
This ends up creating empty directories for stripped modules that don't contain any redirects.

Reply:
Would a simple check for `!m.items.is_empty()` suffice?

Reply:
No. The modules aren't empty, they just don't contain anything that is reexported anywhere else. It's the reason `create_dir_all` was after the `buf.is_empty()` check to make sure directories are only created if there is actually a file to put in it.

Reply:
Would a test for this just involve a `#[doc(hidden)]` module that doesn't have anything re-exported from it? I think i can recreate the check, i just need to make sure i'm doing it right.

Reply:
A simple crate like the following is enough to demonstrate this issue:

```rust
#[doc(hidden)]
pub mod hidden_mod {
    pub mod more_hidden {}
}
```

I don't recommend trying to recreate the logic though. If `create_dir_all` is really that expensive then I suggest adding a `bool` to `Context` to keep track of whether the current directory has been created yet.

Reply:
The repeated creation of directories adds up for big crates like libc or winapi, especially if you make documentation for these on windows, where filesystem operations tend to be more expensive. The problem with just adding a `bool` to `Context` is that `Context::krate` makes a clone of the Context for each item, so it only sees the state of the Context as it was when it had only passed the module itself. Putting it into `SharedContext` is a problem because there's no way to reset the SharedContext when `Context::krate` is done with it. To do it fully properly, you'd need to make it a `HashSet<PathBuf>` in the `SharedContext`, which trades your filesystem operations for a bunch of allocations. Is that a worthwhile tradeoff when checking the directory is relatively cheap?

Reply:
The logic it uses isn't that bad to recreate, unless you're worried about various HashMap lookups. I've force-pushed an update to this commit that should trim out these empty directories.

Reply:
Yeah, adding a `bool` to `Context` doesn't work as well as I'd thought.

The new code misses at least one check: https://github.com/rust-lang/rust/blob/1ed7d41d8849d930f0622eaf54049f66fff0ca2a/src/librustdoc/html/render.rs#L1533
We don't generate pages for items with no name, specifically re-exports. The following creates an empty directory with the new version:

```rust
#[doc(hidden)]
pub mod hidden_mod {
    pub use super::Foo;
}
pub struct Foo;
```

It would be good to see some benchmark numbers for this to see if the added complexity is worth it.

Also, why is this in the `write_shared` function?

Reply:
@retep998 ran procmon on an earlier iteration of this PR yesterday when rendering winapi documentation to verify the number of syscalls per-item went down. For that crate, anything that reduces the number of filesystem accesses per-item is worth it, simply because it has a ton of items being exported. I'll try to get some before/after numbers for this, at least on my own machine.

As for being in `write_shared`... that's where i decided to put it? I think it felt weird to put it into the main `run` function, but that does seem better, now that you point it out. I think i picked `write_shared` just because it was the last thing before the final docs were rendered.  I'll move it into `run`, that's no problem for me.

Reply:
PM 094225 <WindowsBunnyDoesSupportIndexing> misdreavus: 114 seconds in nightly to 107 seconds using semi recent version of your PR

(When they render documentation on winapi)

I've pushed the latest update. I'm going to try to update `htmldocck.py` to allow for checking whether a directory exists. That will have to be a separate PR, though.